src/components/tl/ucp/reduce_scatter/reduce_scatter_ring.c

```diff
@@ -0,0 +1,362 @@
+    int                n_frags, frag;
+    size_t             block_count;
+
+    n_frags = task->reduce_scatter_ring.n_frags;
+    frag    = task->reduce_scatter_ring.frag;
+
+    block_count   = ucc_buffer_block_count(count, size, block);
+    *frag_offset  = ucc_buffer_block_offset(block_count, n_frags, frag);
+    *block_offset = ucc_buffer_block_offset(count, size, block);
+}
+
+static inline ucc_status_t ucc_tl_ucp_test_ring(ucc_tl_ucp_task_t *task)
+{
+    int polls = 0;
+    while (polls++ < task->n_polls) {
+        if (task->send_posted - task->send_completed <= 1 &&
+            task->recv_posted == task->recv_completed) {
+            return UCC_OK;
+        }
+        ucp_worker_progress(TASK_CTX(task)->ucp_worker);
+    }
+    return UCC_INPROGRESS;
+}
+
+static ucc_status_t
+ucc_tl_ucp_reduce_scatter_ring_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t *task     = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_coll_args_t *  args     = &TASK_ARGS(task);
+    ucc_tl_ucp_team_t *team     = TASK_TEAM(task);
+    ucc_rank_t         size     = task->subset.map.ep_num;
+    ucc_rank_t         rank     = task->subset.myrank;
+    void *             sbuf     = args->src.info.buffer;
+    ucc_memory_type_t  mem_type = args->dst.info.mem_type;
+    size_t             count    = args->dst.info.count * size;
+    ucc_datatype_t     dt       = args->dst.info.datatype;
+    size_t             dt_size  = ucc_dt_size(dt);
+    ucc_rank_t         sendto   = (rank + 1) % size;
+    ucc_rank_t         recvfrom = (rank - 1 + size) % size;
+    ucc_rank_t         prevblock, recv_data_from;
+    ucc_status_t       status;
+    size_t max_block_size, block_offset, frag_count, frag_offset, final_offset;
+    int    step, is_avg;
+    void * tmp[3];
+
+    final_offset = 0;
+    if (UCC_IS_INPLACE(*args)) {
+        sbuf = args->dst.info.buffer;
+        count /= size;
+        final_offset =
+            ucc_buffer_block_offset(count, size, UCC_TL_TEAM_RANK(team));
+    }
+
+    sendto   = ucc_ep_map_eval(task->reduce_scatter_ring.inv_map, sendto);
+    recvfrom = ucc_ep_map_eval(task->reduce_scatter_ring.inv_map, recvfrom);
+
+    max_block_size = ucc_buffer_block_count(count, size, 0) * dt_size;
+    tmp[0]         = task->reduce_scatter_ring.scratch;
+    tmp[1]         = PTR_OFFSET(tmp[0], max_block_size);
+    tmp[2]         = PTR_OFFSET(tmp[1], max_block_size);
+
+    if (UCC_INPROGRESS == ucc_tl_ucp_test_ring(task)) {
+        return task->super.super.status;
+    }
+    while (task->recv_posted > 0) {
+        void *reduce_target = tmp[1 + (task->recv_completed - 1) % 2];
+        step                = task->send_posted;
+        prevblock           = (rank - 1 - step + size) % size;
+        prevblock =
+            ucc_ep_map_eval(task->reduce_scatter_ring.inv_map, prevblock);
+        if (task->recv_completed == size - 1) {
+            ucc_assert(prevblock == UCC_TL_TEAM_RANK(team));
```

Comment:
what is the purpose of this assert? first checking prevblock is rank and assign to rank in next line

Reply:
right - was part of debug. removing.

Reply:
correct - removed (part of debug)